Memory map executing test case failed when code coverage enabled in x86_64 platform

**Describe the bug**

In the case we call k_mem_map() to mapping memory a piece of code snippet, then execute it via mapped pointer, it will access wrong Gcov stubbed variable address inside the mapped code. This is due to this code snippet was complied using %rip indirect addressing mode in qemu_x86_64. For qemu_x86, the same code snippet was complied using absolute addressing mode.  

This bug was found when I try to enable the code coverage feature for qemu_x86_64. Please refer to PR https://github.com/zephyrproject-rtos/zephyr/pull/29873

Details descriptions:
If we enable gcov to generate code coverage data,  it will stub some extra code and variables into the original code first in order to record statistical data.

But for one special test case in X86 64 platform, it would cause a unexcepted fault.
See tests/kernel/mem_protect/mem_map/src/main.c , one test case called “test_z_mem_map_exec()”
The error message is here:

START - test_z_mem_map_exec
E: Page fault at address 0xc0021e5e (error code 0x0)
E: Linear address not present in page tables
E: Access violation: supervisor thread not allowed to read
E: PDPTE: not present
E: RAX: 0x000000000014c73f RBX: 0x0000000000131f40 RCX: 0x0000000000130001 RDX: 0x00000000bfff4000
E: RSI: 0x0000000000000000 RDI: 0x000000000014c73f RBP: 0x000000000014c720 RSP: 0x000000000014c720
E:  R8: 0x0000000000000008  R9: 0xffffffffffffffff R10: 0x0000000000000000 R11: 0x0000000000000000
E: R12: 0x0000000000000001 R13: 0x0000000000000000 R14: 0x0000000000000000 R15: 0x0000000000000000
E: RSP: 0x000000000014c720 RFLAGS: 0x0000000000000202 CS: 0x0018 CR3: 0x0000000000178000
E: call trace:
E: RIP: 0x00000000bfff4008
E:      0x00000000001004f2
E:      0x000000000010c64d
E:      0x000000000010c7bb
E:      0x0000000000101053
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x131e80 (ztest_thread)
Caught system error -- reason 0
Unexpected fault during testFAILED: zephyr/CMakeFiles/run
cd /home/mm/zephyrproject/zephyr/build && /home/mm/zephyr-sdk-0.11.4/sysroots/x86_64-pokysdk-linux/usr/bin/qemu-system-x86_64 -m 9 -cpu qemu64,+x2apic -device isa-debug-exit,iobase=0xf4,iosize=0x04 -no-reboot -nographic -net none -pidfile qemu.pid -chardev stdio,id=con,mux=on -serial chardev:con -mon chardev=con,mode=readline -device loader,file=/home/mm/zephyrproject/zephyr/build/zephyr/zephyr-qemu-main.elf -smp cpus=2 -kernel /home/mm/zephyrproject/zephyr/build/zephyr/zephyr-qemu-locore.elf
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/mm/zephyrproject/zephyr/build --target run

Then I tried to diagnose this issue:
In this test case, we want to map memory for using code snippet of  transplanted_function().
And gcov also stub into this function:  (this assembly code comes from zephyr.lst)

0000000000100412 <transplanted_function>:
#ifndef SKIP_EXECUTE_TESTS
static void transplanted_function(bool *executed)
{
  100412:       55                                 push   %rbp
  100413:       48 89 e5                       mov    %rsp,%rbp
  100416:       48 89 7d f8                   mov    %rdi,-0x8(%rbp)
  **10041a:       48 8b 05 4f de 02 00    mov    0x2de4f(%rip),%rax         # 12e270 <__gcov0.transplanted_function>**
  100421:       48 83 c0 01                   add    $0x1,%rax
  100425:       48 89 05 44 de 02 00    mov    %rax,0x2de44(%rip)        # 12e270 <__gcov0.transplanted_function>
        *executed = true;
  10042c:       48 8b 45 f8                    mov    -0x8(%rbp),%rax
  100430:       c6 00 01                        movb   $0x1,(%rax)
}
  100433:       90                                 nop
  100434:       5d                                 pop    %rbp
  100435:       c3                                 retq


As you can see, in x86_64 platform, the stubbed gcov variable address here is indirect addressing mode with %rip.
When we call transplanted_function() directly, the correct addressing should be 2de4f+10041a(rip) is close to 12e270.
But in the memory mapping case, the %rip will be different, and the address of gcov variable here will be incorrect.   
The %rip would be new mapping address base such as bfff4000, So the incorrect addressing is 2de4f+bfff4008(rip) is close to c0021e5e, where is not preset in Page table. 

For x86 platform,  the address of gcov stub here was complied as absolute addressing, it won’t cause problem in this case.  
Below code snippet comes from zephyr.lst of qemu_x86 with coverage. 

00100898 <transplanted_function>:
#ifndef SKIP_EXECUTE_TESTS
static void transplanted_function(bool *executed)
{
  100898:       55                              push   %ebp
  100899:       89 e5                         mov    %esp,%ebp
  **10089b:       a1 18 b6 12 00          mov    0x12b618,%eax**
  1008a0:       8b 15 1c b6 12 00      mov    0x12b61c,%edx
  1008a6:       83 c0 01                     add    $0x1,%eax
  1008a9:       83 d2 00                     adc    $0x0,%edx
  1008ac:       a3 18 b6 12 00           mov    %eax,0x12b618
  1008b1:       89 15 1c b6 12 00      mov    %edx,0x12b61c
        *executed = true;
  1008b7:       8b 45 08                    mov    0x8(%ebp),%eax
  1008ba:       c6 00 01                     movb   $0x1,(%eax)
}
  1008bd:       90                              nop
  1008be:       5d                              pop    %ebp
  1008bf:       c3                               ret

** As far as I know, it seems like if we want to execute code snippet which got by memory mapping, the assembly code which gcov stubbed should using absolute addressing mode.**


**To Reproduce**
Steps to reproduce the behavior:
sanitycheck -T tests/kernel/mem_protect/mem_map/ -p qemu_x86_64 --coverage
or 
1. west build tests/kernel/mem_protect/mem_map/ -b qemu_x86_64 -DCONFIG_COVERAGE=y -DCONFIG_COVERAGE_GCOV=y
2. west build -t run

**Expected behavior**
Tests must pass without timeout errors or freezing

**Impact**
It makes the memory map test case of executing failed when code coverage enabled for x86_64 platform.

**Logs and console output**


**Environment (please complete the following information):**
 - OS: Ubuntu 18.04.3 LTS
 - Toolchain Zephyr SDK zephyr-sdk-0.11.4 and Zephyr SDK zephyr-sdk-0.12.0-beta-2
 - Commit SHA 8f2ba929

**Additional context**

--- a/tests/kernel/mem_protect/mem_map/CMakeLists.txt
+++ b/tests/kernel/mem_protect/mem_map/CMakeLists.txt
@@ -11,3 +11,5 @@ target_include_directories(app PRIVATE
 
 FILE(GLOB app_sources src/*.c)
 target_sources(app PRIVATE ${app_sources})
+
+zephyr_linker_sources(SECTIONS custom-sections.ld)

--- a/tests/kernel/mem_protect/mem_map/custom-sections.ld
+++ b/tests/kernel/mem_protect/mem_map/custom-sections.ld
@@ -1,0 +1,13 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+SECTION_DATA_PROLOGUE(TEST_MEM_MAP,,SUBALIGN(CONFIG_MMU_PAGE_SIZE))
+{
+	/* transplanted_function() must be first */
+	KEEP(*(".test_mem_map.transplanted_function*"));
+	KEEP(*(".test_mem_map.*"));
+
+} GROUP_DATA_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
+
+PROVIDE(__test_mem_map_start = LOADADDR(TEST_MEM_MAP));
+PROVIDE(__test_mem_map_size = SIZEOF(TEST_MEM_MAP));
+PROVIDE(__test_mem_map_end = __test_mem_map_start + __test_mem_map_size);

--- a/tests/kernel/mem_protect/mem_map/prj_x86_64_coverage_exec.conf
+++ b/tests/kernel/mem_protect/mem_map/prj_x86_64_coverage_exec.conf
@@ -1,0 +1,6 @@
+CONFIG_ZTEST=y
+
+# Enable large memory model to avoid relative addressing
+# so the execution test can run with the function at
+# new location.
+CONFIG_COMPILER_OPT="-mcmodel=large"

--- a/tests/kernel/mem_protect/mem_map/src/main.c
+++ b/tests/kernel/mem_protect/mem_map/src/main.c
@@ -13,15 +13,6 @@
 #define SKIP_EXECUTE_TESTS
 #endif
 
-/* Skip the memory map executing case when coverage enabled in x86_64,
- * because it will crash due to incorrect address accessing for gcov variables.
- * See issue#30434 for more details.
- */
-#if defined(CONFIG_X86_64) && defined(CONFIG_COVERAGE)
-#define SKIP_EXECUTE_TESTS
-#endif
-
-
 #define BASE_FLAGS	(K_MEM_CACHE_WB)
 volatile bool expect_fault;
 

--- a/tests/kernel/mem_protect/mem_map/src/main.c
+++ b/tests/kernel/mem_protect/mem_map/src/main.c
@@ -6,6 +6,7 @@
 
 #include <ztest.h>
 #include <sys/mem_manage.h>
+#include <toolchain.h>
 
 /* 32-bit IA32 page tables have no mechanism to restrict execution */
 #if defined(CONFIG_X86) && !defined(CONFIG_X86_64) && !defined(CONFIG_X86_PAE)
@@ -87,6 +88,11 @@ void test_z_phys_map_rw(void)
 }
 
 #ifndef SKIP_EXECUTE_TESTS
+extern char __test_mem_map_start[];
+extern char __test_mem_map_end[];
+extern char __test_mem_map_size[];
+
+__in_section_unique(test_mem_map) __used
 static void transplanted_function(bool *executed)
 {
 	*executed = true;
@@ -99,29 +105,30 @@ static void transplanted_function(bool *executed)
  */
 void test_z_phys_map_exec(void)
 {
-	uint8_t *mapped_rw, *mapped_exec, *mapped_ro;
+	uint8_t *mapped_exec, *mapped_ro;
 	bool executed = false;
 	void (*func)(bool *executed);
 
 	expect_fault = false;
 
-	/* Map with write permissions and copy the function into the page */
-	z_phys_map(&mapped_rw, (uintptr_t)test_page,
-		   sizeof(test_page), BASE_FLAGS | K_MEM_PERM_RW);
-
-	memcpy(mapped_rw, (void *)&transplanted_function, CONFIG_MMU_PAGE_SIZE);
+	/*
+	 * Need to reference the function or else linker would
+	 * garbage collected it.
+	 */
+	func = transplanted_function;
 
 	/* Now map with execution enabled and try to run the copied fn */
-	z_phys_map(&mapped_exec, (uintptr_t)test_page,
-		   sizeof(test_page), BASE_FLAGS | K_MEM_PERM_EXEC);
+	z_phys_map(&mapped_exec, (uintptr_t)__test_mem_map_start,
+		   (uintptr_t)__test_mem_map_size,
+		   BASE_FLAGS | K_MEM_PERM_EXEC);
 
 	func = (void (*)(bool *executed))mapped_exec;
 	func(&executed);
 	zassert_true(executed, "function did not execute");
 
 	/* Now map without execution and execution should now fail */
-	z_phys_map(&mapped_ro, (uintptr_t)test_page,
-		   sizeof(test_page), BASE_FLAGS);
+	z_phys_map(&mapped_ro, (uintptr_t)__test_mem_map_start,
+		   (uintptr_t)__test_mem_map_size, BASE_FLAGS);
 
 	func = (void (*)(bool *executed))mapped_ro;
 	expect_fault = true;

--- a/tests/kernel/mem_protect/mem_map/testcase.yaml
+++ b/tests/kernel/mem_protect/mem_map/testcase.yaml
@@ -2,3 +2,4 @@ tests:
   kernel.memory_protection.mem_map:
     tags: kernel mmu ignore_faults
     filter: CONFIG_MMU
+    extra_sections: _TRANSPLANTED_FUNC

--- a/tests/kernel/mem_protect/mem_map/testcase.yaml
+++ b/tests/kernel/mem_protect/mem_map/testcase.yaml
@@ -1,5 +1,20 @@
 tests:
   kernel.memory_protection.mem_map:
     tags: kernel mmu ignore_faults
-    filter: CONFIG_MMU
+    filter: CONFIG_MMU and not CONFIG_X86_64
     extra_sections: _TRANSPLANTED_FUNC
+    platform_exclude: qemu_x86_64
+  kernel.memory_protection.mem_map.x86_64:
+    tags: kernel mmu ignore_faults
+    filter: CONFIG_MMU and CONFIG_X86_64 and not CONFIG_COVERAGE
+    extra_sections: _TRANSPLANTED_FUNC
+  kernel.memory_protection.mem_map.x86_64.coverage:
+    tags: kernel mmu ignore_faults
+    filter: CONFIG_MMU and CONFIG_X86_64 and CONFIG_COVERAGE
+    extra_sections: _TRANSPLANTED_FUNC
+    extra_args: EXTRA_CFLAGS=-DSKIP_EXECUTE_TESTS
+  kernel.memory_protection.mem_map.x86_64.coverage.exec:
+    tags: kernel mmu ignore_faults
+    filter: CONFIG_MMU and CONFIG_X86_64 and CONFIG_COVERAGE
+    extra_sections: _TRANSPLANTED_FUNC
+    extra_args: CONF_FILE=prj_x86_64_coverage_exec.conf

